Feature/ap configuration from m2

YuriyMasechko opened the pull request:
Agent part of #346 .

Bwl was added as a stub. Need to implement bwls separately and modify test_flows.sh accordingly.

YuriyMasechko committed:
tlvf: Add ACTION_APMANAGER_WIFI_CREDENTIALS_UPDATE_REQUEST

Add ACTION_APMANAGER_WIFI_CREDENTIALS_UPDATE_REQUEST to update wifi
credentials based on M2.

Signed-off-by: Yuriy Masechko <yuriy.masechko@globallogic.com>

YuriyMasechko committed:
bwl: add update_wifi_credentials function

Add bwl function to upate wifi credentials.

Signed-off-by: Yuriy Masechko <yuriy.masechko@globallogic.com>

YuriyMasechko committed:
agent: call bwl update_wifi_credentials based on M2

Call bwl update_wifi_credentials for each M2 received except case
if teardown received with other M2 flags.

Signed-off-by: Yuriy Masechko <yuriy.masechko@globallogic.com>

arnout commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_apmanager.yaml:
```diff
@@ -184,6 +184,23 @@ cACTION_APMANAGER_STEERING_EVENT_AUTH_FAIL_NOTIFICATION:
+  enc_type:
+    _type: char    
+    _length: [ "beerocks::message::WIFI_SECURITY_TYPE_MAX_LENGTH" ]
+  is_backhaul: uint8_t
```
A BSS has 3 possible modes: fronthaul, backhaul, or backhaul+fronthaul combined. So this should be either:
- a bitmap with fronthaul and backhaul bits;
- an enum with 3 values;
- two boolean fields.

For the first two options, the name is_backhaul is not appropriate. In hostapd, it's an enum with 4 values (because there's also the "not EasyMesh" case), and it's called "multi_ap".

morantr replied:
Why not using `eWscVendorExtsubelementBssType`? This will also include the teardown.

YuriyMasechko replied:
Changed to list of configdata.

arnout commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_apmanager.yaml:
```diff
@@ -184,6 +184,23 @@ cACTION_APMANAGER_STEERING_EVENT_AUTH_FAIL_NOTIFICATION:
+    _length: [ "beerocks::message::WIFI_SECURITY_TYPE_MAX_LENGTH" ]
+  enc_type:
+    _type: char    
+    _length: [ "beerocks::message::WIFI_SECURITY_TYPE_MAX_LENGTH" ]
```
These two should be enums instead of freeform strings.

I would even use `WSC::eWscAuth` directly - @tomereli @morantr  what do you think?

morantr replied:
Agree.

tomereli replied:
Agree

arnout commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_apmanager.yaml:
```diff
@@ -184,6 +184,23 @@ cACTION_APMANAGER_STEERING_EVENT_AUTH_FAIL_NOTIFICATION:
+    _type: char    
+    _length: [ "beerocks::message::WIFI_SECURITY_TYPE_MAX_LENGTH" ]
+  is_backhaul: uint8_t
+  teardown: uint8_t
```
Teardown should be a separate action. If we receive a teardown, that should be the only M2 we receive; anything else is an error.

tomereli replied:
I agree but the its another set of messages:/

YuriyMasechko replied:
Changed to list of configdata.

arnout commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_apmanager.yaml:
```diff
@@ -184,6 +184,23 @@ cACTION_APMANAGER_STEERING_EVENT_AUTH_FAIL_NOTIFICATION:
+  _type: class
+  ssid:
+    _type: char    
+    _length: [ "beerocks::message::WIFI_SSID_MAX_LENGTH" ]
```
Maybe it was implemented like that before, but this is wrong: the length should be dynamic. So we should have a separate ssid_length field:
```
  ssid_length:
    _type: uint8_t
  ssid:
    _type: char
    _length: [ ssid_length ]
```

morantr replied:
But add property `_length_var: True` to the `ssid length`:
```
  ssid_length:
    _type: uint8_t
    _length_var: True
  ssid:
    _type: char
    _length: [ ssid_length ]
```

tomereli replied:
I don't think this is required. Check out the var length lists in M2.

YuriyMasechko replied:
Changed to list of configdata.

arnout commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_apmanager.yaml:
```diff
@@ -184,6 +184,23 @@ cACTION_APMANAGER_STEERING_EVENT_AUTH_FAIL_NOTIFICATION:
+    _length: [ "beerocks::message::WIFI_SSID_MAX_LENGTH" ]
+  pass:
+    _type: char    
+    _length: [ "beerocks::message::WIFI_PASS_MAX_LENGTH" ]
```
Same here, length should be dynamic.

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4622,6 +4662,32 @@ bool slave_thread::handle_autoconfiguration_wsc(Socket *sd, ieee1905_1::CmduMess
                    << (backhaul ? " backhaul" : "") << (teardown ? " teardown" : "");
+
+        if (!fronthaul && !backhaul && !teardown)
+            continue;
```
This should be logged as an error. But the `continue` part is good.

In addition, I would also add a check that there's either a single teardown, or one or more front/backhaul M2s. So it should not be allowed that there is no M2 at all, or that a teardown is combined with anything else. Of course, the question is what to do in that situation: tear down? partially configure? I would say to do a teardown. So the check is simple: if teardown is true, then m2_list should have size 1.

YuriyMasechko replied:
According new design error cases are ignored.

arnout reviewed (changes requested):
It would be useful if there was also an action "reset_wifi_credentials" before the loop over the M2s, and "apply_wifi_credentials" afterwards, because we want to apply all credentials in one shot. Actually, it would be even better if all credentials were passed in a single action, and the bwl call has a vector of credentials structs.

Actually, we already have bss_info_conf_t in db.h, so maybe we could move that (except for the operating_class list) to some common header and use it in bwl as well.

morantr commented on common/beerocks/bwl/include/bwl/ap_wlan_hal.h:
```diff
@@ -157,6 +157,23 @@ class ap_wlan_hal : public virtual base_wlan_hal {
+
+    virtual bool set_wifi_credentials(const std::string &ssid, const std::string &pass,
+                                      const std::string &auth, const std::string &enc,
+                                      uint8_t is_backhaul, uint8_t teardown) = 0;
```
As @arnout has said, `auth` and `enc` should be enums, and `is_backhauk` and `teardown` should be bitfield or enum (cpp enum is valid if its value moves between min to max value).

YuriyMasechko replied:
Changed to bssinfo list.

morantr commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -69,6 +69,42 @@ static bwl::WiFiSec platform_to_bwl_security(const std::string &sec)
+    default:
+        return "INVALID";
+    }
+}
```
I think it will better be on son_wireless_utils.

YuriyMasechko replied:
Moved.

tomereli reviewed (changes requested):
Comments inline

tomereli commented:
![](https://github.trello.services/images/mini-trello-icon.png) [autoconfig M2 support in agent](https://trello.com/c/hPNsRYNF/15-autoconfig-m2-support-in-agent)

YuriyMasechko committed:
common: move bss_info_conf_t to son wireless utils

Move bss_info_conf_t to wireless utils to have possibility
to pass wireless configuration to bwl as list of bss infos.

Signed-off-by: Yuriy Masechko <yuriy.masechko@globallogic.com>

YuriyMasechko committed:
comomon: add functions to convert wsc wifi info to bwl

Add function to convert WSC wifi settings to BWL one.

Signed-off-by: Yuriy Masechko <yuriy.masechko@globallogic.com>

tomereli commented on common/beerocks/bwl/dummy/ap_wlan_hal_dummy.cpp:
```diff
@@ -172,6 +172,22 @@ bool ap_wlan_hal_dummy::sta_bss_steer(const std::string &mac, const std::string
+                   << " encr_type: "
+                   << son::wireless_utils::wsc_to_bwl_encryption(bss_info_conf.encryption_type)
+                   << " network_key: " << bss_info_conf.network_key << " bss_type: "
+                   << son::wireless_utils::wsc_to_bwl_bss_type(bss_info_conf.bss_type);
```
Perhaps we can store the credentials in some class member, and use it in `refresh_radio_info` which is sent to the controller. This way even in Dummy mode in Linux containers we will see the credentials in `bml_conn_map`.
Based on this, we can add/update the corresponding test flow to verify that the correct VAP credentials were applied.
This can be done in a followup PR, but I think it *should* be done.

arnout replied:
I added this comment to #343.

tomereli commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -635,6 +635,13 @@ bool ap_wlan_hal_dwpal::sta_bss_steer(const std::string &mac, const std::string
+bool ap_wlan_hal_dwpal::update_vap_credentials(
+    std::list<son::wireless_utils::sBssInfoConf> &bss_info_conf_list)
+{
+    //TODO Implement #346
```
Better to have a full link to the issue, in the future we might move to gitlab so perhaps the issues numbers won't be the same (nitpick)

tomereli commented on agent/src/beerocks/slave/ap_manager_thread.cpp:
```diff
@@ -811,6 +812,41 @@ bool ap_manager_thread::handle_cmdu(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_
+                config_data.multiap_attr().subelement_value);
+
+            bss_info_conf_list.push_back(bss_info_conf);
+        }
```
Just wanted to say that this looks much better now, good work with the yaml :)

tomereli commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3577,7 +3577,7 @@ bool slave_thread::autoconfig_wsc_authenticate(std::shared_ptr<ieee1905_1::tlvWs
     std::shared_ptr<ieee1905_1::tlvWscM2> m2, uint8_t authkey[32], uint8_t keywrapkey[16],
     std::string &ssid, sMacAddr &bssid, WSC::eWscAuth &auth_type, WSC::eWscEncr &encr_type,
-    bool &backhaul, bool &fronthaul, bool &teardown)
+    bool &backhaul, bool &fronthaul, bool &teardown, std::shared_ptr<WSC::cConfigData> &credentials)
```
Don't we have the backhaul, fronthaul and teardown bits as part of `cConfigData`? If so, now that you pass the whole `configData` I don't see a reason to pass another 3 booleans.

arnout replied:
> Don't we have the backhaul, fronthaul and teardown bits as part of `cConfigData`?

In fact, *all* the credentials are in cConfigData; we don't need ssid, bssid, auth_type, encr_type either. However, the backhaul, fronthaul and teardown bits are a bit difficult to retrieve since they're part of the vendor extension attribute.

I originally proposed to use cConfigData, but I'm thinking that maybe we should define a new, cleaner structure instead...

Anyway, I'm OK with using cConfigData for now.

arnout replied:
Just to clarify: I agree with @tomereli that it doesn't make sense to return anything other than `credentials`.

Another thing, however: instead of a shared_ptr, it should be a reference. A reference makes it clear that the function will not keep the object alive after it returns. Cfr. [CppCoreGuildeline R.30](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rr-smartptrparam).

YuriyMasechko replied:
Removed ssid, bssid, auth_type, encr_type. Backhaul, fronthaul and teardown left as is because they should be parsed as bit map.
It should be shared_ptr because inside tlvf it is shared ptr. Could change yaml to have it as reference.

tomereli commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3823,8 +3838,21 @@ bool slave_thread::handle_autoconfiguration_wsc(Socket *sd, ieee1905_1::CmduMess
+
+        if (teardown && m2_list.size() > 1) {
+            // Teardown is not compatible with any other bit in the M2.
+            LOG(DEBUG) << "Teardwon combined with other M2, ignoring all";
```
should be `LOG(ERROR)`

YuriyMasechko replied:
Done.

tomereli reviewed (commented):
Looks good in general, although I have some more comments I think could be useful as part of this PR.
Especially regarding the saving of the credentials in dummy mode and using them in `refresh_radio_info` so they will show in the `bml_conn_map` in the controller. This however can be done later.

Not approving yet so Arnout has a chance to review. Please fix the remaining issues and I think we should merge this today or on Sunday since this is a blocker for the autoconfig task.

arnout reviewed (approved):
Excellent splitting into commits! Every commit is easy to review on its own, and the tests succeed after every commit.

I agree with @tomereli's comments and I added one small additional comment. So I'm approving already.

tomereli reviewed (approved):
LGTM

morantr commented on common/beerocks/bwl/dummy/ap_wlan_hal_dummy.cpp:
```diff
@@ -173,6 +173,22 @@ bool ap_wlan_hal_dummy::sta_bss_steer(const std::string &mac, const std::string
+bool ap_wlan_hal_dummy::update_vap_credentials(
+    std::list<son::wireless_utils::sBssInfoConf> &bss_info_conf_list)
+{
+    for (auto bss_info_conf : bss_info_conf_list) {
```
please use const reference
`const auto &bss_info_conf`

I guess you can use const on the function signature ass well.

morantr reviewed (approved):
Small comment, nothing important.

tomereli commented:
Please rebase and merge
